py-polars/polars/utils/udfs.py

```diff
@@ -165,22 +168,59 @@ def _bytecode_to_expression(
     return " ".join(expr for _, expr in exprs)
 
 
-def _can_rewrite_as_expression(ops: list[ByteCodeInfo], apply_target: str) -> bool:
+def _can_rewrite_as_expression(
+    ops: list[ByteCodeInfo],
+    apply_target: str,
+    param_name: str,
+    function: Callable[[Any], Any],
+) -> bool:
     """
     Determine if bytecode indicates only simple binary ops and/or comparisons.
 
     Note that `lambda x: x` is inefficient, but we ignore it because it is not
     guaranteed that using the equivalent bare constant value will return the
     same output. (Hopefully nobody is writing lambdas like that anyway...)
     """
+    if len(ops) < 3:
+        return False
     simple_ops = _SIMPLE_FRAME_OPS if apply_target == "frame" else _SIMPLE_EXPR_OPS
-    if bool(ops) and all(op[0] in simple_ops for op in ops) and len(ops) >= 3:
-        return (
-            # can (currently) only handle logical 'and'/'or' if they are not mixed
-            len({_LOGICAL_OPCODES[op[0]] for op in ops if op[0] in _LOGICAL_OPCODES})
-            <= 1
-        )
-    return False
+    logical_opcodes = set()
+    for i, op in enumerate(ops):
+        if op[0] == "BINARY_SUBSCR":
+            if (
+                i < 2
+                or ops[i - 1][1] != param_name
+                or ops[i - 2][0] not in ("LOAD_GLOBAL", "LOAD_DEREF")
+            ):
+                return False
+            import inspect
+
+            dict_name = ops[i - 2][1]
+            frameinfo = inspect.stack(0)[-1]
+            for frameinfo in inspect.stack(0):
+                if dict_name in frameinfo.frame.f_locals:
+                    if function in list(
+                        frameinfo.frame.f_locals.values()
+                    ) and isinstance(frameinfo.frame.f_locals[dict_name], dict):
+                        return True
+                elif dict_name in frameinfo.frame.f_globals:
+                    if function in list(
+                        frameinfo.frame.f_locals.values()
+                    ) and isinstance(frameinfo.frame.f_globals[dict_name], dict):
+                        return True
```

Comment:
right, so this is possible! I should've known better than to doubt the possibilities of what Python offers

gonna check this more carefully later

it adds about `.0001` seconds runtime, which isn't as bad as I was expecting - and it would only be happening for the `map_dict` case, not for every single `apply` call, so not too bad at all